# H10 - DJ Ango

Hanna Gröndahl

## Järjestelmätiedot

- Rauta: Dell Latitude 7290
- Raudan OS: Windows 10 Pro (10.0.19045)
- Hypervisor: Oracle VM Virtual Box 7.0.6
- Virtuaalikoneen OS: Debian GNU/Linux 11
- Kernel: Linux 5.10.0-20-amd64

## a) Kannattavaa.

Aloitin Djangon asentamisen ja käytön päivittämällä pakettilistauksen `sudo apt-get update` ja asentamalla virtualenvin `sudo apt-get -y install virtualenv`. Virtualenvin tarkoituksena on luoda erillinen Python-ympäristö. Komennolla `virtualenv --system-site-packages -p python3 env/` loin Pythonin käyttöä varten virtuaaliympäristön. --system-site-packages mahdollistaa myös virtualenvin ulkopuolisten (apt-)pakettien käytön. Python3 asentaa Python3-tulkin. Siiryin virtuaaliympäristöön komennolla `source env/bin/activate`. Kehotteen alkuun tuli `(env)`, mikä osoittaa, että toimin virtualenvissä.

![env-prompt](https://user-images.githubusercontent.com/122886984/221207404-159875bf-f82d-4922-b7a4-6d4cfaa2bd68.png)

Django täytyy asentaa pip:llä, joka on Pythonin pakettinhallintatyökalu. Tarkistin, että pip sijaitsee virtualenvissä komennolla `which pip`, ja komento tulosti hakemiston, jossa pip sijaitsee:

    /home/hanna/env/bin/pip

Kaikki oli sen osalta siis kunnossa. Loin microlla requirements.txt-tiedoston, jonka pohjalta tulisin asentamaan pip:llä tarvittavat paketit. Kirjoitin tiedostoon "django".

![requirements](https://user-images.githubusercontent.com/122886984/221207461-0d16469f-e4f6-45e9-aa5f-a2d24837b6c3.png)

Tein asennuksen komennolla `pip install -r requirements.txt`. Tarkistin asennetun version komennolla ´django-admin --version`. Tulostus oli 4.1.7, mikä kuulosti hyvältä, sillä se on uudempi kuin ohjeissa ollut 4.0.2. Django oli nyt asennettu.

Aloitin Djangon käytön luomalla hannapp-projektin komennolla `django-admin startproject hannapp`. Siirryin hannapp-hakemistoon `cd hannapp` ja käynnistin testipalvelimen komennolla `./manage.py runserver`. Komentokehote näytti virheilmoituksena, että migraatioita ei ole otettu käyttöön.

![migrate](https://user-images.githubusercontent.com/122886984/221207658-ae1d1b09-71aa-48c9-9d9c-f0648762cde4.png)

Testipalvelin käynnistyi osoitteeseen http://127.0.0.1:8000/. Katsoin selaimella, että asennus onnistui. Django antoi hyväksyntänsä asennukselle.

![runserver](https://user-images.githubusercontent.com/122886984/221207744-c5f3a187-b425-47e4-9a88-cd070a58ebd5.png)

Sen jälkeen poistuin serveriltä ja tein pyydetyt migraatiot.

![migrate2](https://user-images.githubusercontent.com/122886984/221207760-b44329f5-a8fe-4681-8f80-a85c827c8ab1.png)

Käyttääkseni Djangoa admin-oikeuksilla selaimella lisäsin urlin perään /admin/. Ensin sain "Sivustoon ei saada yhteyttä" -virheilmoituksen, mutta laitettuani testipalvelimen takaisin päälle `./manage.py runserver` sain admin-kirjautumisnäkymän. Tällä kertaa ei tullut virheilmoituksiakaan palvelimen käynnistyksessä, sillä olin juuri tehnyt migraation. Poistuin palvelimelta ja loin pääkäyttäjän `./manage.py createsuperuser`. Täytin pyydetyt tiedot, käynnistin palvelimen ja kokeilin kirjautumista. Kirjautuminen onnistui.

![admin](https://user-images.githubusercontent.com/122886984/221207893-1414c333-0042-46d6-9f93-dcc6fe491a0c.png)

Käytin graafista käyttöliittymää selaimessa luodakseni uuden käyttäjän. Loin manu-käyttäjän, ja annoin manulle staff- ja superuser-statukset. Kirjauduin hannalla ulos ja kokeilin kirjautua manulla. Pääsin sisään. Kokeilin oikeuksia luomalla uuden käyttäjän rajatuilla oikeuksilla. Kirjauduin ulos ja kokeilin kirjautua tällä uudella käyttäjällä. Käyttäjä ei päässyt sisään, koska se ei ole staff-käyttäjä.

![staff](https://user-images.githubusercontent.com/122886984/221207934-ae141bf7-a765-4b11-97c5-54d993840d6c.png)

Loin asiakastietokannan komennolla `./manage.py startapp crm`. Lisäsin sen asennettuihin sovelluksiin komennolla `micro hannapp/settings.py` ja lisäämällä crm:n INSTALLED_APPS -luetteloon. 

![installed](https://user-images.githubusercontent.com/122886984/221207964-92a16258-64c2-4656-81f2-40bda3dcd4d7.png)

Tein models.py-tiedoston `micro crm/models.py`, jotta Django voi automaattisesti luoda sen pohjalta tietokantoja. Loin sinne Cat-luokan, johon liittyy name-kenttä. Tein migraation ja lisäsin sen admin-rekisteriin luomalla ensin admin.py-tiedoston `micro crm/admin.py` ja lisäämällä mallin sinne.

![adminpy](https://user-images.githubusercontent.com/122886984/221207989-33d16516-87db-4d8a-a6de-3e9197af142f.png)

Käynnistin testipalvelimen `./manage.py runserver`, jotta voin tarkistaa, tuliko tietokanta näkyviin. Kirjauduin admin-näkymän kautta ja siellä oli CRM:n alla Cats-välilehti. Ihmettelin, että ohjeessa luokka oli nimetty yksikkönä eikä monikkona kuten taulut yleensä, mutta Django näköjään fiksusti muutti taulun nimen monikkoon ja hyödynsi yksikkömuotoa muualla. Lisäsin tietokantaan pari kissaa.

![addcat](https://user-images.githubusercontent.com/122886984/221208042-0163b324-d874-4ce0-aaf9-993a8ad99494.png)

Jotta listauksessa näkyisi kissojen nimet eikä Cat object (1) ja (2), lisäsin models.py-tiedostoon funktion, joka palauttaa Cat-luokan objektin nimen name-kentän tiedon mukaan. Käynnistin testipalvelimen `./manage.py runserver`, ja päivitin tietokannan selaimessa. Nyt kissat näkyivät siellä käytännöllisesti nimien mukaan eikä numeroituina kissaobjekteina. 

![kissanimet](https://user-images.githubusercontent.com/122886984/221208068-13ae1506-6dab-445a-8363-d48a43bdf919.png)

## Lähteet

- Karvinen, T. 2016. Django 4 Instant Customer Database Tutorial. https://terokarvinen.com/2022/django-instant-crm-tutorial/. Luettu 24.2.2023.
- Karvinen, T. 2023. Linux-palvelimet-kurssin luento 17.2.2023.
